README.md
# wppExplorerBayesMig  

[![Travis-CI Build Status](https://travis-ci.org/PPgp/wppExplorerBayesMig.svg?branch=master)](https://travis-ci.org/PPgp/wppExplorerBayesMig)

Shiny-based R package to view data from the PPgp/wpp2015BayesMig package, which is a modified version of the World Population Prospects (WPP) published by the United Nations Population Division (package wpp2015). In this version of the data, uncertainty about migration was taken into account.

#####Reference:

Azose, J.J., &#352;ev&#269;&#237;kov&#225;, H., Raftery, A.E. (2016): <a href="http://www.pnas.org/content/113/23/6460.full">Probabilistic population projections with migration uncertainty</a>.  <i>Proceedings of the National Academy of Sciences</i> 113:6460–6465.

To start a session, from R do 
```R
library(wppExplorerBayesMig)
wpp.explore()
```
This will open the interface in your default web browser. To end the session, press the Esc key.

The package also offers various functions to retrieve the WPP indicators in R as data frames. See the help file for `?wpp.indicator`.




ChangeLog
0.2-1 (01/06/2017)
------
Bug fix in Rosling chart (affects sex- and age-specific indicators)
	
0.2-0 (11/1/2016)
------
Changed shinytheme.

Changed color scheme for maps of indicators with negative values (thanks to Jon Azose).

In histogram: Values on the x-axis rounded (thanks to Jon Azose). 

Code updated to make a use of the shiny grid system.

Tables made scrollable.

0.1-0 (9/16/2016)
------
Initial version.


tests/testthat.R
library(testthat)
library(wppExplorerBayesMig)

test_check("wppExplorerBayesMig")


tests/testthat/test-indicators.R
context('Indicators')

set.year <- function(wpp.year) {
	if(!get.wpp.year()==wpp.year)
		set.wpp.year(wpp.year)
}

test_that('fertility age profile has the right dimension', {
	prof <- wppExplorerBayesMig:::get.age.profile.fert(2000, c('FR', 'AF'))
	expect_true(all(dim(prof) == c(2*7, 4))) # 2 countries by 7 age groups
	prof <- wppExplorerBayesMig:::get.age.profile.fert(2020, 'GE')
	expect_true(all(dim(prof) == c(7, 4)))
	expect_true(all(prof$charcode == 'GE'))
})

test_that('age-specific fertility rate has the right value', {
	ind <- wpp.by.year(wpp.by.country(wpp.indicator('fertage', age="20-24"), 'FR'), 2015)$value
	data(tfr, package="wpp2015")
	data(percentASFR, package="wpp2015")
	tfrFR <- subset(tfr, country_code==250)["2010-2015"]
	asfrFR <- subset(percentASFR, country_code==250 & age=="20-24")["2010-2015"]
	expect_true(ind == tfrFR*asfrFR/100)
})


test_that('mortality values for high ages come out correctly', {
	set.year(2015)
	mx <- wpp.by.year(wpp.by.country(wpp.indicator('mortagesex', sex="M", age="100+"), 'FR'), 2015)$value
	expect_true(mx > 0.51)
	mx <- wpp.by.year(wpp.by.country(wpp.indicator('mortagesex', sex="M", age="100+"), 'FI'), 2015)$value
	expect_true(mx > 0.5 & mx < 0.51)
	mx2 <- wpp.by.year(wpp.by.country(wpp.indicator('mortagesex', sex="M", age="100"), 'FI'), 2015)$value
	expect_true(all.equal(mx, mx2))
	mx <- wpp.by.year(wpp.by.country(wpp.indicator('mortagesex', sex="F", age="110"), 'FI'), 2015)$value
	expect_true(mx > 1.2)
})

test_that('migration rate has the right value', {
	ind <- wpp.by.year(wpp.by.country(wpp.indicator('migrate'), 'AE'), 2005)$value
	data(pop, package="wpp2015BayesMig")
	data(migration, package="wpp2015BayesMig")
	popUA <- sum(subset(pop, country_code==784)[c("2000", "2005")])/2
	migUA <- subset(migration, country_code == 784)["2000-2005"]
	expect_true(ind == migUA*200/popUA)
})

R/global.R
utils::globalVariables("wpp.data.env")
get.indicator.choices <- function() {
	ind.names <- c('Total Fertility Rate', 'Female Life Expectancy', 'Male Life Expectancy', 
					'Total Population', 'Female Population', 'Male Population', 
					'Net Migration', 'Net Migration Rate', 
					'Sex Ratio at Birth', 'Median Age', 'Mean Age at Childbearing', 'Mean Age of Women in Childbearing Ages',
					'Total Dependency Ratio', 'Child Dependency Ratio', 'Old-age Dependency Ratio','Potential Support Ratio',
					'Mean Annual Population Growth',
					'Population by sex and age', 'Mortality Rate by sex and age', 'Age-specific Fertility Rate', 'Percent Age-specific Fertility')
	ind.def <- c('', '', '',
				'Total population in thousands', 'Female population in thousands', 'Male population in thousands', 
				'Net migration counts in thousands per 5 years', 'Annual net migration rate (per thousand population; denominator is approx. average population)',
				'Ratio of male to female', '', 'Mean age of mothers at the birth of their children', 'Mean age of women that are in childbearing ages',
				'Ratio of population age 0-14 and 65+ to population age 15-64',
				'Ratio of population age 0-14 to population age 15-64', 
				'Ratio of population age 65+ to population age 15-64', 
				'Ratio of population age 15-64 to population age 65+', 
				'log(P_t/P_{t-1})/5', 'Population in thousands', '', '', '')
	funcs <- c('fert', 'leF', 'leM', 'tpop', 'tpopF', 'tpopM', 'mig', 'migrate',
				'sexratio', 'medage', 'meanagechbear', 'meanageinchbearage',
				'tdratio', 'chdratio', 'oadratio', 'psratio',
				'popgrowth',
				'popagesex', 'mortagesex', 'fertage', 'pfertage')
	# if a new indicator is added, change also the condition in ui.R for displaying age-specific stuff
	l <- length(ind.names)
	ini <- rep(FALSE, l)
	ind.df <- data.frame(by.age=ini, no.age.sum=ini, sum.in.table=ini, low.high=ini, half.child=ini, prob.ci=ini, mid.years=ini,
							digits=rep(1, l), has.negatives=ini) 
	rownames(ind.df) <- funcs
	ind.df[c('popagesex', 'mortagesex', 'fertage', 'pfertage'), 'by.age'] <- TRUE  # display sex and age menu
	ind.df[c('mortagesex','fertage'), 'no.age.sum'] <- TRUE                        # don't allow multiple age- and sex-selection
	ind.df[c('tpop', 'tpopF', 'tpopM', 'mig','popagesex'), 'sum.in.table'] <- TRUE # show sum in the trend table
	ind.df[c('fert', 'leF', 'leM', 'tpop', 'tpopF', 'tpopM', 'popagesex', 'mig', 'migrate'), 'low.high'] <- TRUE       # has uncertainty
	ind.df[c('fert', 'tpop', 'tpopF', 'tpopM', 'popagesex'), 'half.child'] <- TRUE     # has half.child variant
	ind.df[c('fert', 'leF', 'leM', 'mig', 'sexratio', 'mortagesex', 'fertage', 'pfertage'), 'mid.years'] <- TRUE # use mid years in slider (not implemented)
	ind.df[c('tpop', 'tpopF', 'tpopM','popagesex', 'mig'), 'digits'] <- 0                 # number of digits the histogram
	ind.df[c('sexratio', 'popgrowth', 'mortagesex'), 'digits'] <- 4
	ind.df['fertage', 'digits'] <- 3
	ind.df[c('tdratio', 'chdratio', 'oadratio'), 'digits'] <- 2
	ind.df[c('mig', 'migrate', 'popgrowth'), 'has.negatives'] <- TRUE
	
	structure(
		as.character(1:length(ind.names)),
		names = ind.names,
		definition = ind.def,
		settings = ind.df
	)
}


assign("wpp.data.env", new.env(), envir=parent.env(environment())
	#envir = .GlobalEnv
	)
data('iso3166', envir=wpp.data.env)
wpp.data.env$indicators <- get.indicator.choices()
wpp.data.env$package <- c("wpp2015BayesMig", "wpp2015")
# Filter out non-used countries
do.call('data', list("popM", package=wpp.data.env$package, envir=wpp.data.env))
wpp.data.env$iso3166 <- wpp.data.env$iso3166[is.element(wpp.data.env$iso3166$uncode, wpp.data.env$popM$country_code),]
#wpp.data.env$sim.dir <- '/mnt/shinydata/apps/wppExplorerBayesMig/simulation' # stat rstudio shiny server
wpp.data.env$sim.dir <- '/Users/hana/bayespop/R/Pop/wpp2015/sim07292015' # grafton
#wpp.data.env$sim.dir <- '/Users/hana/bayespop/R/Pop/migration/wpp2015/mig1T07152016_15-34' # brno



inst/explore/ui.R
library(shinythemes)

geochartPrereqs <- tagList(
  tags$head(
    tags$script(src="https://www.google.com/jsapi"),
    tags$script(src="geochart.js")
  )
)

geochart <- function(id, options=list()) {
  tags$div(id=id, class="shiny-geochart-output", `data-options`=RJSONIO::toJSON(options))
}

googleLineChart <- function(id, options=list()) {
  tags$div(id=id, class="google-linechart-output", `data-options`=RJSONIO::toJSON(options))
}

googleHistogram <- function(id, options=list()) {
  tags$div(id=id, class="google-histogram-output", `data-options`=RJSONIO::toJSON(options))
}

 
shinyUI(
fluidPage(theme = shinytheme("yeti"),
  titlePanel(paste("WPP", wppExplorerBayesMig:::get.wpp.year(), "Explorer with Migration Uncertainty")),
 sidebarLayout(
  sidebarPanel(
    shinyjs::useShinyjs(),
    geochartPrereqs,
    tags$head(
		tags$style(type='text/css', ".span4 { max-width: 270px; }")
	),
    uiOutput('yearUI'),
    hr(),
    selectInput('indicator', h5('Indicator:'), wppExplorerBayesMig:::wpp.data.env$indicators),
    conditionalPanel(condition=paste("input.indicator >", sum(attr(wppExplorerBayesMig:::wpp.data.env$indicators, "settings")$by.age == FALSE)),
    	tags$head(tags$style(type="text/css", "#selagesmult { height: 150px; width: 85px}"),
    			  tags$style(type="text/css", "#selages { width: 85px}"),
    			  tags$style(type="text/css", "#indsexmult { height: 55px; width: 95px}"),
    			  tags$style(type="text/css", "#indsex { width: 95px}")
    	),
    	fluidRow(
			column(4, offset=2, uiOutput('sexselection')),
    		column(3, offset=1, uiOutput('ageselection'))
    	)
    ),
    htmlOutput('indicatorDesc'),
    hr(),
    selectInput('uncertainty', h5('Uncertainty:'), structure(as.character(1:3), names=c('80%', '95%', '+-1/2child')), 
    			multiple=TRUE, selected=1),
    textOutput('uncertaintyNote'),
    hr(),
    HTML("<p><small><b>Data Source:</b> Based on estimates for 1950-2015 from <a href='http://esa.un.org/unpd/wpp' target='_blank'>World Population Prospects 2015</a>. Probabilistic projections including migration uncertainty based on <a href='http://www.pnas.org/content/113/23/6460.full' target='_blank'>Azose et al. (2016, PNAS)</a>. Note: These are not official <a href='https://rstudio.stat.washington.edu/shiny/wppExplorer/inst/explore/' target='_blank'> UN projections</a>.</small></p><p><small>&copy; Hana &#352;ev&#269;&#237;kov&#225;, <a href='https://www.csss.washington.edu' target='_blank'>CSSS</a>, University of Washington; <a href='http://bayespop.csss.washington.edu' target='_blank'>project website</a></small></p>"),
width=3
  ),
  mainPanel(
    shinyjs::useShinyjs(),
    tabsetPanel(
      tabPanel('Map',
      	 fluidRow(
		column(6, checkboxInput('normalizeMapAndCountryPlot', 'Fixed scale over time', TRUE))
		),
		fluidRow(
		column(6, offset=5, textOutput('mapyear'))
			),		
		hr(),
		geochart('map'),
		#htmlOutput('mapgvis'),
		hr(),
		conditionalPanel(condition='input.map_selection',
				plotOutput('countryPlot', height='300px'))
      ),
      tabPanel('Sortable Data', 
      	fluidRow(
		column(6, checkboxInput('includeAggr2', 'Include Aggregations', FALSE))
		),
	fluidRow(
		column(6, offset=5, textOutput('year2'))
		),
		hr(),
      	DT::dataTableOutput('stable')
      ),
      tabPanel('Trends & Pyramids',
  		tags$head(
			tags$style(type="text/css", "#seltcountries { height: 450px}"),
			tags$style(type="text/css", "#trendstable { overflow-x: scroll}")
			),
			fluidPage(
				fluidRow(
					fluidRow(HTML("<br>")),
					column(3, uiOutput('cselection')),
				  	column(9, 
				  		tabsetPanel(
				  				tabPanel('Trends', 
				  					plotOutput('probtrends', 
				  							click = "probtrends_values", hover = "probtrends_values", 
				  							dblclick = "probtrends_zoom_reset", 
				  							brush = brushOpts(id = "probtrends_zoom", resetOnNew = TRUE)),
				  					flowLayout(
										checkboxInput('trend.logscale', 'Log scale', FALSE),
										textOutput("probtrends_selected")
										)
				  					),
				  				tabPanel('Age Profile', 
				  					googleLineChart('age.profileM', options=list(height=200)),
				  					googleLineChart('age.profileF', options=list(height=200)),
				  					checkboxInput('aprofile.logscale', 'Log scale', FALSE)),
				  				tabPanel('Pyramids', 
				  					plotOutput('pyramids', click = "pyramid_values", hover = "pyramid_values", 
				  						dblclick = "pyramid_zoom_reset",
				  						brush = brushOpts(id = "pyramid_zoom", resetOnNew = TRUE)
									),
				  					flowLayout(
				  						checkboxInput('proppyramids', 'Pyramid of proportions', FALSE),
				  						textOutput("pyramid_selected")
									)
				  				), type="pills"
				  			)
				  		)
 					),
 				fluidRow(
 					column(12, textOutput('trendstabletitle'),
 						   tableOutput('trendstable'))
 					)
 				)
 		),
 	tabPanel('Histogram',
 		flowLayout(
      		checkboxInput('fiXscaleHist', 'Fixed x-axis over time', FALSE)
      	),
      	hr(),
      	#plotOutput('hist')
      	htmlOutput('ghist')
    ),
	tabPanel('Probability Calc',
		tags$head(
			tags$style(HTML('#probcalculate{background-color:lightblue}')),
			tags$style(type="text/css", "#probcalc_result { overflow-x: scroll}")
			),
		fluidPage(
				fluidRow(HTML("<br>")),
				fluidRow(
					column(12, 
						HTML("<h5><font color='#d95f0e'><b>Probability that population in the selected country will be larger/smaller than the threshold:</b></font></h5>"))
				),
				fluidRow(
					column(12, HTML("<p></p>"))
				),
				fluidRow(
					column(3, uiOutput('cselection_single')),
					column(3, textInput("probcalc_threshold", "Threshold (in thousands):")),
					column(3, radioButtons("probcalc_direction", label="", choices=list(larger=1, smaller=0), selected=1, inline=TRUE))
				),
				fluidRow(
					column(3, actionButton('probcalculate', 'Calculate Probability'))
				),
				fluidRow(
					column(12, HTML("<p></p><p></p>"))
				),
				fluidRow(
					column(12, HTML("<h5><font color='#d95f0e'><b>Results:</b></font></h5>"),
						textOutput('probcalctabletitle'))
				),
				fluidRow(
					column(12, tableOutput("probcalc_result"))
				)
		)	
	),   
    tabPanel('Rosling Chart',
		htmlOutput('graphgvis'),
		HTML("<br>"),
		fluidRow(column(1,"")),
		fluidRow(
			column(3, offset=1, textOutput('AddIndicatorText')),
			column(1, actionButton("AddIndicator", "Add indicator"))
			)
      ),
      tabPanel("Help",
      	includeHTML("README.html")
      ) 
      ) #end tabsetPanel
  ) #end mainPanel
)))
